files/zh-cn/web/api/history_api/working_with_the_history_api/index.md

```diff
@@ -1,105 +1,192 @@
 
-如果用户再次点击**返回**，URL 将变为 `https://mozilla.org/foo.html`，并且文档将得到一个 `popstate` 事件，这次是一个 `null` 状态对象。在这里，返回也不会改变文档的内容，尽管文档在收到 `popstate` 事件后可能会手动更新其内容。
+这些 API 的主要目的是支持像{{Glossary("SPA", "单页应用")}}这样的网站，它们使用 JavaScript API（如 {{domxref("fetch()")}}）来更新页面的新内容，而不是加载整个新页面。
 
-### pushState() 方法
+## 单页应用和会话历史记录
 
-`pushState()` 需要三个参数：一个**状态对象**；一个**标题**（目前忽略）；以及一个（可选）**URL**。
+传统上，网站是以页面集合的形式实现的。当用户通过点击链接浏览网站的不同部分时，浏览器每次都会加载一个全新的页面。
 
-让我们详细地解释一下这三个参数
+虽然这对许多网站来说很好，但也有一些缺点：
 
-- **状态对象**
-  - : 状态对象是一个 JavaScript 对象，它与由 `pushState()` 创建的新历史条目相关。每当用户浏览到新的状态时，一个 `popstate` 事件被触发，该事件的 `state` 属性包含历史条目的状态对象的副本。状态对象可以是任何可以被序列化的东西。因为 Firefox 将状态对象保存在用户的磁盘上，以便在用户重新启动浏览器后恢复，所以我们对状态对象的序列化表示法规定了 640k 字符的大小限制。如果你传递给 `pushState()` 的状态对象的序列化表示大于这个大小，该方法将抛出一个异常。如果你需要比这更多的空间，我们鼓励你使用 `sessionStorage` 和/或 `localStorage`。
-- **标题**
-  - : [除了 Safari，所有的浏览器目前都忽略了这个参数](https://github.com/whatwg/html/issues/2174)，尽管它们在未来可能会使用这个参数。在这里传递空字符串应该是安全的，可以避免将来对该方法的修改。另外，你也可以为你要移动的状态传入一个简短的标题。
-- **URL**
-  - : 新的历史条目的 URL 是由这个参数给出的。注意，在调用 `pushState()` 后，浏览器不会尝试加载这个 URL，但它可能会在以后尝试加载这个 URL，例如在用户重新启动浏览器后。新的 URL 不需要是绝对的；如果它是相对的，它就会相对于当前的 URL 进行解析。新的 URL 必须与当前的 URL 同源；否则，`pushState()` 将抛出一个异常。这个参数是可选的；如果没有指定它，它将被设置为文档的当前 URL。
+- 如果只需要更新页面的一部分，每次加载整个页面的效率会很低。
+- 跨页面导航时很难保持应用程序状态
 
-在某种意义上，调用 `pushState()` 类似于设置 `window.location = "#foo"`，因为两者都将创建并激活与当前文档相关的另一个历史条目。
+由于这些原因，{{Glossary("SPA", "单页应用")}}（SPA）是一种流行的 web 应用程序模式，在这种模式中，网站由一个页面组成，当用户点击链接时，页面会：
 
-但是 `pushState()` 有几条优势：
+1. 阻止加载新页面的默认行为
+2. {{domxref("fetch()", "获取", "", "nocode")}}显示新内容
+3. 用新内容更新页面
 
-- 新的 URL 可以是与当前 URL 同源的任何 URL。相比之下，只有当你只修改锚点值时，设置 `window.location` 使你保持在同一个 {{ domxref("document") }}。
-- 如果你不愿意，你不必改变 URL。相反，设置 `window.location = "#foo";` 只在当前锚点值不是 `#foo` 的情况下创建一个新的历史条目。
-- 你可以将任意的数据与你的新历史条目联系起来。使用基于哈希的方法，你需要将所有的相关数据编码成一个短字符串。
-- 如果 `title` 随后被浏览器使用，这个数据可以被利用（而锚点值会独立于每个历史条目）。
+例如：
 
-请注意，`pushState()` 永远不会触发 `hashchange` 事件，即使新的 URL 与旧的 URL 只在锚点值上有所不同。
+```js
+document.addEventListener("click", async (event) => {
+  const creature = event.target.getAttribute("data-creature");
+  if (creature) {
+    // 阻止新页面加载
+    event.preventDefault();
+    try {
+      // 获取新内容
+      const response = await fetch(`creatures/${creature}.json`);
+      const json = await response.json();
+      // 用新内容更新页面
+      displayContent(json);
+    } catch (err) {
+      console.error(err);
+    }
+  }
+});
+```
 
-在其他文档中，它创建了一个具有 `null` 命名空间 URI 的元素。
+在该点击处理程序中，如果链接包含数据属性 `"data-creature"`，我们就会使用该属性的值来获取包含页面新内容的 JSON 文件。
 
-### replaceState() 方法
+JSON 文件可能如下所示：
 
-`history.replaceState()` 的操作与 `history.pushState()` 完全一样，只是 `replaceState()` 修改了当前的历史条目，而不是创建一个新的。注意，这并不妨碍在全局浏览器历史中创建一个新条目。
+```json
+{
+  "description": "秃鹫并不“秃”。",
+  "image": {
+    "src": "images/eagle.jpg",
+    "alt": "一只秃鹫"
+  },
+  "name": "秃鹫"
```

Comment:
Let's use the untranslated content (or this example is inconsistent with the live example)